src/traits/block_storage.rs

```diff
@@ -0,0 +1,37 @@
+//! Abstract storage type for storing DA proposals and VID shares
+//!
+//! This modules provides the [`BlockStorage`] trait.
+//!
+
+use async_trait::async_trait;
+use snafu::Snafu;
+
+use crate::{
+    data::{DAProposal, VidDisperse},
+    message::Proposal,
+};
+
+use super::node_implementation::NodeType;
+
+#[derive(Snafu, Debug)]
+#[snafu(visibility(pub))]
+pub enum BlockStorageError {
+    #[snafu(display("Failed to store the proposal due to an internal error"))]
+    StoreError,
+
+    #[snafu(display("Failed to retrieve the proposal due to an internal error"))]
+    RetrieveError,
+}
+
+#[derive(Clone, Debug)]
+pub enum ProposalType<TYPES: NodeType> {
```

Comment:
Nit: I prefer using `anyhow::Error` over ad hoc error types for each trait, like this. Especially for traits, because we can't necessarily predict at the trait definition what types of errors the implementation may encounter, and `anyhow` allows the implementation to return arbitrary descriptive errors which aids in debugging.

However if there is some convention in this repo of using enums for errors, I defer to that.

Reply:
I also prefer anyhow, let me check with the team if anyone has a strong opinion.